confluent_kafka/schema_registry/json_schema.py

```diff
@@ -0,0 +1,266 @@
+        elif not self._auto_register and subject not in self._known_subjects:
+            registered_schema = self._registry.lookup_schema(subject,
+                                                             self._schema)
+            self._schema_id = registered_schema.schema_id
+            self._known_subjects.add(subject)
+
+        if self._to_dict is not None:
+            value = self._to_dict(ctx, obj)
+        else:
+            value = obj
+
+        try:
+            validate(instance=value, schema=self._parsed_schema)
+        except ValidationError as ve:
+            raise SerializationError(ve.message)
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record to the rest of the buffer
+            json.dump(value, fo, value)
+
+            return fo.getvalue()
+
+
+class JsonDeserializer(Deserializer):
+    """
+    JsonDeserializer decodes bytes written in the Schema Registry
+    JSON format to an object.
+
+    Args:
+        schema_registry_client (SchemaRegistryClient): Confluent Schema Registry
+            client instance.
+
+        schema_str (str): JSON schema declaration.
+
+        from_dict (callable, optional): Callable(SerializationContext, dict) -> object.
+            Converts dict to an instance of some object.
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """
+    __slots__ = ['_schema', '_registry', '_from_dict']
+
+    def __init__(self, schema_registry_client, schema_str, from_dict=None):
+        self._registry = schema_registry_client
+        self._schema = json.loads(schema_str)
+
+        if from_dict is not None and not callable(from_dict):
+            raise ValueError("from_dict must be callable with the signature"
+                             " from_dict(SerializationContext, dict) -> object")
+
+        self._from_dict = from_dict
+
+    def __call__(self, ctx, value):
+        """
+        Deserializes Schema Registry response to JSON object literal(dict)
+
+        Args:
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+            value (bytes): bytes
+
+        Returns:
+            dict: Deserialized JSON
+
+        Raises:
+            SerializerError: If response payload and expected Message type
+            differ.
+
+        """
+        if len(value) <= 5:
+            raise SerializationError("message is too small to decode {}".format(value))
+
+        with _ContextStringIO(value) as payload:
+            magic, schema_id = struct.unpack('>bI', payload.read(5))
+            if magic != _MAGIC_BYTE:
+                raise SerializationError(
+                    "message does not start with magic byte")
+
+            # JSON documents are self-describing; no need to query schema
```

Comment:
But the schema dictates what fields are required, and which fields should be ignored (not included in schema).
If the Deserializer doesn't check it then the application needs to, which sort of defeats the purpose of a schema aware deserializer.

Reply:
you are right added with a test

Reply:
@mhowlett Just to verify; is the .NET client validating received JSON with the schema?

Reply:
No, but this is consistent with the AvroDeserializer and ProtobufDeserializer which is a 1:1 relationship as was requested when the generic serdes API was proposed.

Reply:
in .net there is no explicit validation against any schema on deserialization for json or protobuf. you would most want to do this against the reader schema (the schema supplied to the deserializer). i should ideally be doing that. validation against the writer schema seems overly paranoid. 

although i do no explicit validation against the schema, in .net the schema is defined implicitly through attributes on the class, and the json library will do some validation due to that. but we support two types of attributes - those from the JSON library and those from the JSON schema library, and the latter aren't validated against (easy to add without breaking API change).

what i don't know the most about is the need for a schema registry instance. you might want a schema_registry client instance for two reasons. validate against writer schema. maybe some paranoid people want to turn this on. auto-register reader schema, if unknown to SR. maybe some people want that. I don't know. I don't think i'd use it. in .net, confusion about this is not an issue as we can add it later without breaking the API. in python that seems less possible. Overall i'm comfortable with the approach as it stands (document the sr param is ignored). Not ideal, but probably not a big deal.